sklearn/linear_model/_ridge.py

```diff
@@ -1898,38 +1902,42 @@ def _sparse_multidot_diag(self, X, A, X_mean, sqrt_sw):
     def _eigen_decompose_gram(self, X, y, sqrt_sw):
         """Eigendecomposition of X.X^T, used when n_samples <= n_features."""
         # if X is dense it has already been centered in preprocessing
+        xp, is_array_api = get_namespace(X)
         K, X_mean = self._compute_gram(X, sqrt_sw)
         if self.fit_intercept:
             # to emulate centering X with sample weights,
             # ie removing the weighted average, we add a column
             # containing the square roots of the sample weights.
             # by centering, it is orthogonal to the other columns
-            K += np.outer(sqrt_sw, sqrt_sw)
-        eigvals, Q = linalg.eigh(K)
-        QT_y = np.dot(Q.T, y)
+            outer = xp.linalg.outer if is_array_api else np.outer
+            K += outer(sqrt_sw, sqrt_sw)
+        eigvals, Q = xp.linalg.eigh(K)
+        QT_y = Q.T @ y
         return X_mean, eigvals, Q, QT_y
 
     def _solve_eigen_gram(self, alpha, y, sqrt_sw, X_mean, eigvals, Q, QT_y):
         """Compute dual coefficients and diagonal of G^-1.
 
         Used when we have a decomposition of X.X^T (n_samples <= n_features).
         """
+        xp, is_array_api = get_namespace(eigvals)
         w = 1.0 / (eigvals + alpha)
         if self.fit_intercept:
             # the vector containing the square roots of the sample weights (1
             # when no sample weights) is the eigenvector of XX^T which
             # corresponds to the intercept; we cancel the regularization on
             # this dimension. the corresponding eigenvalue is
             # sum(sample_weight).
-            normalized_sw = sqrt_sw / np.linalg.norm(sqrt_sw)
+            norm = xp.linalg.vector_norm if is_array_api else np.linalg.norm
```

Comment:
I think one could write a lightweight narwhals version for array API

Reply:
We already have `array_api_compat` and our own numpy wrapper that we could leverage to help us do that.

Reply:
There is a new `array-api-extra` project that has started to implement missing utilities on top of what is standardized in the spec and therefore implemented in the `array-api-compat` project: 

https://data-apis.org/array-api-extra/